### DISCUSSION TITLE
Blocks are not translatable

### DISCUSSION THREAD

**Discussion by moonbyt3** [woocommerce/woocommerce#44596] [Checkout Flow]
I'm having a huge problem. No matter what I do I can't translate cart. It's half translated to Bosnian, but when I attempt to translate strings in wc-blocks, nothing happens. They seem to be registered with localization function but when translation is applied, they are not read when blocks are being rendered. This is so big issue, I can't believe it went to production.

**Comment by sennevdp** [comment]
I am having the same issue; it's impossible to translate the new Checkout and Cart blocks. When using a translation plugins such as Loco Translate, all strings appear to be correctly translated (to Dutch in my case). However, on the Cart and Checkout pages, most strings are still shown in English.

**Comment by ollidoetsch** [comment]
Same issue here! It's been 6 months since this post and there have been a few others who have described the same problem. Is there really no solution yet?

**Comment by senadir** [comment]
Hey @moonbyt3 @sennevdp @ollidoetsch 

I can confirm Cart/Checkout is translatable fine with a plugin that supports JSON files, including Loco Translate and Say What? plugins.

If you're modifying a po/mo file directly, you need to build the JSON files.

**Comment by aegirt** [comment]
I'm encountering the same issue where the Checkout summary block is not translating using the po/mo files. The billing and address fields in the checkout form are being translated correctly. I've already spent several hours trying to solve this problem and it's quite frustrating. I've installed WooCommerce 9.2.3 and also tried version 9.3 rc. Please address and fix this issue. For your information, I'm translating to Icelandic.

**Comment by MartinBjerg** [comment]
Same problem for me with Danish translation. In contact information theres a mix between danish and english, in shopping cart at local pick up, a mix between english and danish again. It looks very amateurish, customers dont like it and flees.

I have spent extremly many hour trying to solve this, but its like files dont excist, please fix this issue fast.

**Reply by aegirt** [reply]
Hi Martin, 

I found out that the Checkout and Cart blocks are done using JavaScript and have separate translation files in JSON format. My solution was to use the Loco Translate plugin for translations using Google API and that generated the JSON files needed and fixed the problems for me.

Hope this information helps you.

**Comment by senadir** [comment]
Hey! Just to confirm, are you seeing that blocks are not translatable with core languages? or are you trying a translation language, and it's not being applied?

**Comment by MartinBjerg** [comment]
Hello, I an doing a translation in WooCommerce plugin for my language, even
trying to edit the PO files, but its not apply to the page.

man. 16. sep. 2024 08.32 skrev Seghir Nadir ***@***.***>:

**Reply by senadir** [reply]
PO files won't work as they need to be built into JSON files. I suggest you use a plugin/tool that support building into JSON files, or get a POT file that contains reference to said JSON files.

**Comment by MartinBjerg** [comment]
Thanks for your reply, wich plugin would do the job for this?

man. 16. sep. 2024 08.56 skrev Seghir Nadir ***@***.***>:

**Reply by senadir** [reply]
If you're looking to edit simple strings, ["Say What?"](https://wordpress.org/plugins/say-what/) is a good choice.

**Comment by dcardosods** [comment]
I'm experience the same problem with untranslated strings in my checkout page (Bulgarian language). This used to work before migrating to the block based checkout because I had translated the missing translations using Loco plugin.

Now, after extensive debugging, I found the root cause and a workaround.

**Root cause:** WooCommerce block checkout uses React/JavaScript (`wp.i18n.__()`) which loads translations from JSON files in `wp-content/languages/plugins/woocommerce-{locale}-{md5_hash}.json`, not from classic `.po`/`.mo` files. Loco Translate generates its own JSON files in `wp-content/languages/loco/plugins/` and has a `LoadHelper` that hooks into `load_script_translation_file` / `load_script_translations` filters to merge them at runtime. However, on WordPress 7.0.x, this merge mechanism does not work — the Loco-generated JSON files are simply never merged into the loaded translations.

**Additional finding:** Some Bulgarian translations, e.g. ["Place Order" ](https://translate.wordpress.org/projects/wp-plugins/woocommerce/stable/bg/default/?filters%5Bterm%5D=Place+Order&filters%5Bcase_sensitive%5D=yes&filters%5Bterm_scope%5D=scope_any&filters%5Bstatus%5D=current_or_waiting_or_fuzzy_or_untranslated&filters%5Buser_login%5D=&filter=Apply+Filters&sort%5Bby%5D=priority&sort%5Bhow%5D=desc)actually EXISTS on translate.wordpress.org for the stable WooCommerce project, but are in **"waiting" status** (not approved yet). GlotPress only exports **approved ("current")** translations to the JSON files that WordPress downloads. So even though the translation is submitted, it never makes it to `wp-content/languages/plugins/woocommerce-bg_BG-*.json` files until a translation editor approves it. Maybe this also contributes to the Loco merge not working.

**Workaround:** Add a `load_script_translations` filter in theme's `functions.php` that directly injects missing translations into the JSON data WordPress loads:

```php
add_filter( 'load_script_translations', function( $json, $file, $handle, $domain ) {
    if ( 'woocommerce' !== $domain || ! $json || 'bg_BG' !== determine_locale() ) {
        return $json;
    }
    $data = json_decode( $json, true );
    if ( ! $data || empty( $data['locale_data'] ) ) {
        return $json;
    }
    $key  = key( $data['locale_data'] );
    $msgs = &$data['locale_data'][ $key ];
    $msgs['Place Order']           = [ 'Поръчване' ];
    // add more strings here
    
    return json_encode( $data, JSON_UNESCAPED_UNICODE );
}, 10, 4 );
```

Related issues: #66326

cc: @senadir
